low-level/framework/utils/dmidecode_interface.py

```diff
@@ -0,0 +1,107 @@
+
+    def __init__(self):
+        """init method"""
+        super(Dmidecode, self).__init__()
+
+    def get_cpu_info(self):
+        """Returns online cpus list"""
+        try:
+            cmd = self.DMIDECODE + " -t 4"
+            response, err, _ = SimpleProcess(cmd).run()
+            if err:
+                logger.error("Failed to get list of Online CPUs."
+                             f"ERROR:{err}")
+                return
+            matches = re.findall("Socket Designation:.*|"
+                                 "Status:.*", response.decode())
+
+            cpu_map = {}
+            cpu_list = []
+            cpu_present = cpu_status = None
+            while matches:
+                item = matches.pop(0)
+                if "Designation:" in item:
+                    cpu_str = item.strip().split(": ")[1]
+                    cpu_dig = re.findall('\d+', cpu_str )
+                    cpu_present = cpu_dig.pop(0)
+                    cpu_list.append(int(cpu_present))
+                if "Status:" in item:
+                    cpu_status = item.strip().split(": ")[1]
+                if cpu_present and cpu_status:
+                    cpu_map[cpu_present] = cpu_status
+            logger.debug(f"Mapping of CPU and status:{cpu_map}")
+            online_cpus = []
+            for cpu, status in cpu_map.items():
+                if "Enabled" in status:
+                    online_cpus.append(int(cpu))
+            logger.info(f"Online CPU list:{online_cpus}")
+            return cpu_list, online_cpus
+        except Exception as e:
+            logger.error(f"Failed to get online CPUs info. ERROR:{e}")
+            return
+
+    def get_per_cpu_usage(self):
+        try:
+            cmd = self.DMIDECODE + " -t 4"
+            response, err, _ = SimpleProcess(cmd).run()
+            matches = re.findall("Socket Designation:.*|"
+                                 "Thread Count:.*", response.decode())
+            cpu_thread_map = {}
+            cpu_present = cpu_threads = None
+            per_cpu_usage = []
+            while matches:
+                item = matches.pop(0)
+                if "Designation:" in item:
+                    cpu_str = item.strip().split(": ")[1]
+                    cpu_dig = re.findall('\d+', cpu_str )
+                    cpu_present = cpu_dig.pop(0)
+                if "Thread Count:" in item:
+                    cpu_threads = item.strip().split(": ")[1]
+                if cpu_present and cpu_threads:
+                    cpu_thread_map[cpu_present] = cpu_threads
+            cpu_usage_threads = psutil.cpu_percent(interval=None, percpu=True)
+            for cpu, threads in cpu_thread_map.items():
+                thread_avg = cpu_usage_threads[:int(threads)]
+                cpu_usage_threads = cpu_usage_threads[int(threads):]
+                usage = mean(thread_avg)
+                cpu_usage = round(usage,2)
+                per_cpu_usage.append(cpu_usage)
```

Comment:
can you explain the logic of calculating the usage per cpu ? add some comments also around this code

Reply:
psutil will give the usage for all the cpu threads,
I am taking the mean value of these thread values to get cpu_usage per cpu.
For eg:- 
The thread count for CPU0 is 24, it will calculate mean value of first 24 threads  to get CPU0_usage